Improved performance of the main suggestion function

valin1 opened the pull request:
Fixed the performance of the main suggestions function, to terminate, if the endpoint is already found to be a non-serviceable category, instead of attempting to guess for any other possible categories around the area. (Other suggestions still work as well even after the change) 

- Fixed Sam and Bill's trips through fixing the performance issue
- Added tripid for Sam to use on the phone side to relate the suggestion with the specific trip

AmplabJenkins commented:
Can one of the admins verify this patch?

shankari commented on emission/core/wrapper/suggestion_sys.py:
```diff
@@ -174,25 +185,39 @@ def category_of_business_nominatim(lat, lon):
+            return []
         city = address_dict['city']
         categories = []
+        print("tried worked")
```
remove random print statements; don't want to merge them

shankari commented on emission/core/wrapper/suggestion_sys.py:
```diff
@@ -174,25 +185,39 @@ def category_of_business_nominatim(lat, lon):
+            address, location_is_service = return_address_from_google_nomfile(lat, lon)
+            if (location_is_service == False):
+                return []
+            print("Address")
```
ditto

shankari commented on emission/core/wrapper/suggestion_sys.py:
```diff
@@ -174,25 +185,39 @@ def category_of_business_nominatim(lat, lon):
             categories = []
             possible_bus = match_business_address(address)["businesses"][0]
             possible_categ = possible_bus["categories"]
+            print("Categories")
```
ditto

shankari commented on emission/core/wrapper/suggestion_sys.py:
```diff
@@ -441,6 +466,7 @@ def calculate_yelp_server_suggestion_nominatim(uuid):
     if len(cleaned_trips) == 0:
         return_obj['message'] = 'Suggestions will appear once you start taking trips!'
         return return_obj
+    print(len(cleaned_trips))
```
ditto

shankari commented on emission/core/wrapper/suggestion_sys.py:
```diff
@@ -499,25 +527,25 @@ def calculate_yelp_server_suggestion_nominatim(uuid):
                     #Still looking to see what to return with this message, because currently my latitude and longitudes are stacked together in one string
                     # insert_into_db(tripDict, i, yelp_suggestion_trips, uuid)
-                    return {'message' : message, 'method': 'bike', 'rating': str(ratings_bus[a]), 'businessid': a}
+                    return {'message' : message, 'method': 'bike', 'rating': str(ratings_bus[a]), 'businessid': a, "tripid": str(cleaned_trips.iloc[i]["_id"])}
```
I think you can just return the `tripid` the serializer should handle it

valin1 replied:
I changed it to string, because I believe Sam said the phone uses strings as tripids?

shankari replied:
@valin1 that's what I meant by serializer. The output JSON serializer that converts JSON -> string converts any tripid in the JSON -> string

shankari commented on emission/core/wrapper/suggestion_sys.py:
```diff
@@ -499,25 +527,25 @@ def calculate_yelp_server_suggestion_nominatim(uuid):
                 except ValueError as e:
                     continue
-    return {'message': "Your endpoint has either been a non-serviceable category or a closeby option.",'method': 'public transportation', 'rating': None, 'businessid': None}
+    return {'message': "Your endpoint has either been a non-serviceable category or a closeby option.",'method': 'public transportation', 'rating': None, 'businessid': None, "tripid": None}
```
why is `tripid` `none` here? It's not really an error, right?

valin1 replied:
Changed it.

shankari commented on emission/net/ext_service/geocoder/nominatim.py:
```diff
@@ -166,37 +166,51 @@ def return_address_from_location_google(lat, lon):
         #This try block is for our first 150,000 requests. If we exceed this, use Jack's Token.
         key_string = '&key=' + GOOGLE_MAPS_KEY
         url = base_url + latlng + key_string #Builds the url
+        print(url)
```
pls remove

shankari commented on emission/net/ext_service/geocoder/nominatim.py:
```diff
@@ -166,37 +166,51 @@ def return_address_from_location_google(lat, lon):
             address_comp = cleaned[0]['long_name'] + ' ' + cleaned[1]['short_name']
-            return business_tuple[1], cleaned[3]['short_name'], address_comp
+            #, cleaned[3]['short_name'], address_comp
+            return business_tuple[1], location_is_service
```
I don't think you should change what is returned, you can add on another result though

valin1 replied:
I changed what should be returned because I looked through the code and saw that I never actually used cleaned[3]['short_name'] or address_comp for the rest of my code.

shankari replied:
@valin1 that might be true, but it is very short-term. What if we wanted to merge this into master at some point? What if another group wanted to re-use it? The nominatim file is really shared code, isn't it?

valin1 commented:
Here are the times of testing between the updated and the old function: 

For endpoints that are serviceable and detected to be a service, the time to get a suggestion between both of them are the same, because the function is checking through all the categories for those endpoints. 

Updated function 

CPU times: user 106 ms, sys: 15 ms, total: 121 ms
Wall time: 2.68 s

CPU times: user 92.1 ms, sys: 11 ms, total: 103 ms
Wall time: 2.55 s

CPU times: user 73.6 ms, sys: 9.67 ms, total: 83.2 ms
Wall time: 2.23 s

Old function

CPU times: user 265 ms, sys: 69 ms, total: 334 ms
Wall time: 7.48 s

CPU times: user 313 ms, sys: 121 ms, total: 434 ms
Wall time: 5.96 s

CPU times: user 235 ms, sys: 66 ms, total: 301 ms
Wall time: 6.09 s

Updated Function (Sam’s Trips): 

CPU times: user 555 ms, sys: 47.2 ms, total: 603 ms
Wall time: 11.2 s

CPU times: user 577 ms, sys: 48.6 ms, total: 625 ms
Wall time: 10.3 s

CPU times: user 523 ms, sys: 47.7 ms, total: 570 ms
Wall time: 10.5 s

Old Function (Sam’s Trips:

CPU times: user 595 ms, sys: 56 ms, total: 651 ms
Wall time: 10.7 s

CPU times: user 574 ms, sys: 48.8 ms, total: 623 ms
Wall time: 10.2 s

shankari commented:
@valin1 that's interesting. why are Sam's trips so much different from the rest of the trips?

valin1 commented on emission/core/wrapper/suggestion_sys.py:
```diff
@@ -154,6 +154,17 @@ def match_business_address(address):
+NOMINATIM API: Checks if location given by nominatim call is a residential location
+
+"""
+def is_service_nominatim(business):
```
This was from looking at all the residential places that had the same keywords and for school premises.

valin1 commented:
It took a while to push the commits through as I was mostly checking through all the buggy trips again on the ipython notebook and testing the suggestion_sys code to make sure that everything was working as usual.

shankari commented:
@valin1 two minor changes left. And I really mainly care about the nominatim one.

shankari commented on emission/net/ext_service/geocoder/nominatim.py:
```diff
@@ -168,35 +168,48 @@ def return_address_from_location_google(lat, lon):
+Function that checks if location was a place of service or a residential area. RETURNS TRUE 
+if it is a location of service, FALSE otherwise
+'''
+def isLocationService(address_components):
```
why is this not using `is_service_nominatim` in `nominatim.py`? What is different between the two methods?
